crates/flox-rust-sdk/src/models/environment.rs

```diff
@@ -1,83 +1,764 @@
+            .next()
+            .ok_or(EnvironmentError2::EmptyDotFlox)?
+            .map_err(EnvironmentError2::ReadEnvDir)?;
+
+        let name = EnvironmentName::from_str(&env.file_name().to_string_lossy())
+            .map_err(EnvironmentError2::ParseEnvRef)?;
+
+        Some(Self::open(
+            current_dir,
+            EnvironmentRef::new_from_parts(None, name),
+        ))
+        .transpose()
+    }
+
+    /// Create a new env in a `.flox` directory within a specific path or open it if it exists.
+    ///
+    /// The method creates or opens a `.flox` directory _contained_ within `path`!
+    pub async fn init(
+        path: impl AsRef<Path>,
+        name: EnvironmentName,
+    ) -> Result<Self, EnvironmentError2> {
+        if Self::open(&path, EnvironmentRef::new_from_parts(None, name.clone())).is_ok() {
+            Err(EnvironmentError2::EnvironmentExists)?;
         }
+
+        let env_dir = path.as_ref().join(".flox").join(name.as_ref());
+
+        std::fs::create_dir_all(&env_dir).map_err(EnvironmentError2::InitEnv)?;
+
+        cp_r(env!("FLOX_ENV_TEMPLATE"), &env_dir)
+            .await
+            .map_err(EnvironmentError2::InitEnv)?;
+
+        Self::open(path, EnvironmentRef::new_from_parts(None, name))
     }
 
-    pub fn project(self) -> Option<project::environment::Environment<'flox, Git, ReadOnly<Git>>> {
-        match self {
-            CommonEnvironment::Named(_) => None,
-            CommonEnvironment::Project(p) => Some(p),
+    /// Copy the environment to a sandbox directory given by `path`
+    ///
+    /// Typically within [Flox.temp_dir].
+    /// This implementation tries to stay independent of the [Flox] struct for now.
+    pub async fn modify_in(
+        self,
+        path: impl AsRef<Path>,
+    ) -> Result<Environment<Modify>, EnvironmentError2> {
+        cp_r(&self.path, &path)
+            .await
+            .map_err(EnvironmentError2::MakeSandbox)?;
+
+        Ok(Environment {
+            path: path.as_ref().to_path_buf(),
+            state: Modify { origin: self },
+        })
+    }
+
+    /// Delete the environment
+    ///
+    /// While destructive, no transaction is needed to verify changes.
+    pub fn delete(self) -> Result<(), EnvironmentError2> {
+        std::fs::remove_dir_all(self.path).map_err(EnvironmentError2::DeleteEnvironement)?;
+        Ok(())
+    }
+}
+
+/// Implementations for environments in a "modifiable" state.
+///
+/// Created by [`Environment<Read>::modify_in`].
+/// Provides methods to edit the environment definition file
+/// and commit changes back to the original location of the environemnt.
+impl Environment<Modify> {
+    /// Low level method replacing the definition file content.
+    /// After writing the file, we verify if by trying to build the environment.
+    ///
+    /// This might be deferred to the [`Environment<Modify>::finish`] method in the future.
+    pub async fn set_environment(
```

Comment:
I think replacing the file content and verifying the build are two separate operations that shouldn't be mixed together in the same low-level method. I recommend separate low-level `replace_definition` and `try_build` methods even though they'll be very small methods. This makes testing easier because you can mock out individual operations that either do I/O or aren't what you're concerned with. When in doubt I prefer to provide building blocks and defer to the caller to put them together how they need to (the `flox edit` command handler in this case).

Reply:
Yeah the building within set environment is a it odd, providing that as a separate method or part of `finish` would make more sense (i.e. you shouldn't `finish`/confirm the "transaction" without verifying it can build).
However with the new decalrative spec the build part will be substantially more complex.